### PR TITLE
Move validator attributes to ValidatorConfig

### PR DESCRIPTION
**Pull request by kchulj**
~~Split off from #1197~~
~~The commits will be re-organized once #1197 is complete.~~

~~Relates #816~~
This PR has been closed as the parameters are consensus-critical and will also affect full nodes.

### CODE REVIEW

File: source/agora/test/Base.d [hunk 1]
```diff
@@ -1389,13 +1387,26 @@ public APIManager makeTestNetwork (APIManager : TestAPIManager = TestAPIManager,
         return conf;
     }
 
+    ValidatorConfig makeValidatorConfig (KeyPair node_key)
+    {
+        ValidatorConfig conf =
+        {
+            enabled : true,
+            key_pair : node_key,
+            // preimage_reveal_interval : 1.seconds,  // check revealing frequently
```
Comment (Geod24) [inline source/agora/test/Base.d]: Just remove the comment here and add it in the next commit

File: source/agora/test/Base.d [hunk 2]
```diff
@@ -1421,45 +1432,72 @@ public APIManager makeTestNetwork (APIManager : TestAPIManager = TestAPIManager,
...
+            pair.index % node_configs.length, pair.value[0], node_configs,
+                makeValidatorConfig(pair.value[1])));
+
+    extra_node_configs.drop(test_conf.validators).enumerate(test_conf.validators).each!(
         pair => main_configs ~= makeMainConfig(
             pair.index % node_configs.length, pair.value, node_configs));
 
+    /*extra_node_configs.enumerate.each!(
+        pair => main_configs ~= makeMainConfig(
+            pair.index % node_configs.length, pair.value, node_configs));*/
```
Comment (Geod24) [inline source/agora/test/Base.d]: ?

File: source/agora/test/Base.d [hunk 3]
```diff
@@ -1421,45 +1432,72 @@ public APIManager makeTestNetwork (APIManager : TestAPIManager = TestAPIManager,
...
+
     auto gen_block = makeGenesisBlock(
-        node_configs
-            .filter!(conf => conf.is_validator)
+        validator_configs
+            // .filter!(conf => conf.validator.enabled)
```
Comment (Geod24) [inline source/agora/test/Base.d]: Remove ?

**Review by Geod24 — CHANGES_REQUESTED** [review]
Unfortunately you can't quite split commit this way. Remember, commits should be atomic: It means that on their own, they should compile and pass test, which is clearly not the case here. I can see that the first, second, and last commits are related.

**Comment by kchulj** [comment]
@Geod24 Since the first and second commit rely on each other, do you suggest that I combine these?
Once this one is merged, then I would make the last commits.

**Comment by Geod24** [comment]
Yes. But the last commit is also needed, because without it, the integration tests won't pass. A good litmus test is: "can this commit be alone in it's own PR ?".

File: source/agora/test/Base.d [hunk 4]
```diff
@@ -1421,45 +1432,72 @@ public APIManager makeTestNetwork (APIManager : TestAPIManager = TestAPIManager,
...
     // n2 <- n1 <- n2 and n1 <- o1 and n2 <- o2, (o = outsider).
     // otherwise it will be [n1 <-> n2] <- o1 and [n1 <-> n2] <- o2
-    extra_node_configs.enumerate.each!(
+
+    extra_node_configs.zip(outsider_validator_keys).enumerate
+        .each!(pair => main_configs ~= makeMainConfig(
+            pair.index % node_configs.length, pair.value[0], node_configs,
+                makeValidatorConfig(pair.value[1])));
+
+    extra_node_configs.drop(test_conf.validators).enumerate(test_conf.validators).each!(
```
Comment (linked0) [inline source/agora/test/Base.d]: I guess it's a very clever code but makes the overall code complex. So how about making another variable like `fullnode_configs` for nodes that are not validators. This idea could make the code redundant but more readable.

**Comment by kchulj** [comment]
Okay, I will complete #1197 before moving on here.

File: source/agora/test/Base.d (hunk 4 above)
Comment (kchulj) [reply source/agora/test/Base.d]: Thanks for your suggestion.
The change has been made in #1197.

File: source/agora/common/Config.d [hunk 5]
```diff
@@ -327,49 +327,32 @@ private Config parseConfigImpl (ref const CommandLine cmdln, Node root)
...
-            max_retries : max_retries,
-            timeout : timeout,
-            data_dir : data_dir,
-            validator_cycle : validator_cycle,
-            max_quorum_nodes : max_quorum_nodes,
-            quorum_threshold : quorum_threshold,
-            quorum_shuffle_interval : quorum_shuffle_interval,
-            preimage_reveal_interval : preimage_reveal_interval,
+        genesis_block:
+            opt!(string, "node", "genesis_block")(cmdln, node),
```
Comment (kchulj) [inline source/agora/common/Config.d:332]: Can this be one line?

**Comment by Geod24** [comment]
As discussed, this is a different issue. Those parameters are consensus-critical and will also affect full nodes.

**Comment by AndrejMitrovic** [comment]
> Those parameters are consensus-critical and will also affect full nodes.

`genesis_start_time` and `validator_cycle` definitely belong to both node types.

However I think `max_quorum_nodes`, `quorum_threshold`, `quorum_shuffle_interval`, and `preimage_reveal_interval` can be moved to ValidatorConfig because full nodes don't have to care about the quorums, they only care about the enrolled validator set and the preimages.

**Pull request closed without merging**